### ISSUE TITLE
`--sync-usage-file` Throws Runtime Error: Invalid Memory Address or Nil Pointer Dereference

### ISSUE THREAD

**Issue by AErmie** [infracost/infracost#1013]
When attempting to use the `--sync-usage-file` command, the following runtime error is thrown. 

**InfraCost execution command used:**
```yaml
docker run --rm \
-e INFRACOST_API_KEY=$(InfraCost_API_KEY) \
-e AWS_ACCESS_KEY_ID=$(AccessKeyID_DEV) \
-e AWS_SECRET_ACCESS_KEY=$(SecretAccessKey_DEV) \
--volume $(System.DefaultWorkingDirectory)/InfraCostReport:/InfraCostReport/ \
--volume $(System.DefaultWorkingDirectory)/pipelines/infracost-usage.yml:/InfraCostReport/infracost-usage.yml \
--volume $(System.DefaultWorkingDirectory):/code/ \
infracost/infracost breakdown --sync-usage-file --usage-file InfraCostReport/infracost-usage.yml \
--path /code/terraform/TerraformPlan.json
```

**Runtime Error:**
```cmd
Warning: Ignoring sync-usage-file as no usage-file is specified.

Detected Terraform plan JSON file at /code/terraform/TerraformPlan.json

  ⠋ Syncing usage data from cloud 
Error: An unexpected error occurred

runtime error: invalid memory address or nil pointer dereference
goroutine 1 [running]:
runtime/debug.Stack(0x0, 0x0, 0x0)
	/usr/local/go/src/runtime/debug/stack.go:24 +0x9f
main.handleUnexpectedErr(0xc00016f2c0, 0x1089ca0, 0x1b61a50)
	/app/cmd/infracost/main.go:180 +0x45
main.main.func1(0xc0011adf48, 0xc00016f2c0, 0xc000080120)
	/app/cmd/infracost/main.go:42 +0xa5
panic(0x1089ca0, 0x1b61a50)
	/usr/local/go/src/runtime/panic.go:965 +0x1b9
main.remediateUsage(0xc00016f2c0, 0xc0000ad050, 0x0)
	/app/cmd/infracost/run.go:255 +0x37
main.runMain(0xc000210780, 0xc00016f2c0, 0xc0001d9900, 0x0)
	/app/cmd/infracost/run.go:112 +0x9b4
main.breakdownCmd.func1(0xc000210780, 0xc0002009f0, 0x0, 0x3, 0x0, 0x0)
	/app/cmd/infracost/breakdown.go:41 +0x1c5
github.com/spf13/cobra.(*Command).execute(0xc000210780, 0xc0002009c0, 0x3, 0x3, 0xc000210780, 0xc0002009c0)
	/go/pkg/mod/github.com/spf13/cobra@v1.1.3/command.go:852 +0x472
github.com/spf13/cobra.(*Command).ExecuteC(0xc0001ed680, 0xc0001ed680, 0xc00016f2c0, 0x0)
	/go/pkg/mod/github.com/spf13/cobra@v1.1.3/command.go:960 +0x375
github.com/spf13/cobra.(*Command).Execute(...)
	/go/pkg/mod/github.com/spf13/cobra@v1.1.3/command.go:897
main.main()
	/app/cmd/infracost/main.go:55 +0x105

Environment:
Infracost v0.9.8
```

Also, the [Generate Usage File](https://www.infracost.io/docs/usage_based_resources#1-generate-usage-file) documentation wording implies that if there is no actual usage file provided, InfraCost will "generate a new usage file" itself. 

If this is accurate, there is no indication in the documentation where that newly generated usage file is created (ie. which directory), and what its default name is.

**Comment by aliscott** [comment]
Thanks for reporting this @AErmie. I'm able to reproduce this if I pass `--sync-usage-file` without a `--usage-file` flag specified, but not when I specify a `--usage-file` flag. Is this what you are seeing as well or can you also reproduce this when using both the `--sync-usage-file` and `--usage-file` flags together?

I've pushed a potential fix for this issue in https://github.com/infracost/infracost/pull/1014.

> there is no indication in the documentation where that newly generated usage file is created (ie. which directory), and what its default name is.

The newly generated usage file is created in the location specified by `--usage-file`. I've updated the docs as per below - let me know if you think this will help future people:

> Use the `--sync-usage-file` option to generate a new usage file or update an existing one. You must specify the location of the new or existing usage file using the `--usage-file` flag.

**Comment by AErmie** [comment]
Hey @aliscott, I get the issue when using both `--sync-usage-file` and `--usage-file` flags together. 

I think I (and perhaps the documentation), need some clarity on each flag's use-case. For example, if you use `--sync-usage-file` should it create a new infracost-usage.yml file? If yes, are we supposed to specify the directory to use via the `--usage-file`? If we don't use the `--usage-file` flag, where is the file created, and with what default name? 

Next, if I use the `--usage-file` flag, I understand that will help to provide a more accurate estimate. But, when using it with the `--sync-usage-file` flag, if I point to a pre-existing usage file, that file should be updated. 

When I use both flags together (and only specify a directory instead of an explicit usage file), I get the following error:
> Error: Error creating usage file: open InfraCostReport/: is a directory

When I use both flags together (and also specify an explicit usage file), I get the following error:
> Error: Error creating usage file: open InfraCostReport/infracost-usage.yml: no such file or directory

But, if we check the contents of the InfraCostReport directory, the file is clearly there...

```CMD
Error: Error creating usage file: open InfraCostReport/infracost-usage.yml: no such file or directory
total 44
drwxr-xr-x  2 vsts docker  4096 Oct  4 14:36 .
drwxr-xr-x 15 vsts docker  4096 Oct  4 14:36 ..
-rw-r--r--  1 vsts docker 33230 Oct  4 14:36 Transcoding-Breakdown.json
-rwxr-xr-x  1 root root       0 Oct  4 14:36 infracost-usage.yml
```

**Comment by aliscott** [comment]
Thanks @AErmie! We are currently iterating on the usage file syncing so this is great timing to get this feedback. I'm going to tag @alikhajeh1 and @tim775 here as well if they have ideas about how we can update the docs or interface to improve it based on this.

> For example, if you use --sync-usage-file should it create a new infracost-usage.yml file?

`--sync-usage-file` will not do anything if `--usage-file /path/to/usage-file.yml` is not specified. So on it's own this does not create a new file, and will show the following warning:

```
Warning: Ignoring sync-usage-file as no usage-file is specified.
```

If you pass `--sync-usage-file --usage-file /path/to/usage-file.yml` and `/path/to/usage-file.yml` does not exist then it will create that file. If `/path/to/usage-file.yml` does exist then it will update that file.

> Next, if I use the --usage-file flag, I understand that will help to provide a more accurate estimate. But, when using it with the --sync-usage-file flag, if I point to a pre-existing usage file, that file should be updated.

The `--sync-usage-file` updates the usage file so it has default values for any resources you have. You can then edit the usage file to add in usage estimates for those resources. Once you've added your own estimates, rerunning with `--sync-usage-file` will only add default values for any new resources you add.

> When I use both flags together (and only specify a directory instead of an explicit usage file), I get the following error:

`--usage-file` expects a path to a file not a directory which is why it shows this error.

> When I use both flags together (and also specify an explicit usage file), I get the following error:

I wonder if this is anything to do with the current working directory in the docker container. Does adding a full path like `--usage-file /InfraCostReport/infracost-usage.yml` work?

**Comment by AErmie** [comment]
Hey @aliscott, specifying a full path like you mentioned, seems to have worked. For everyone's reference, this is the pattern that worked for me:

```yaml
docker run --rm \
-e INFRACOST_API_KEY=$(InfraCost_API_KEY) \
-e AWS_ACCESS_KEY_ID=$(AccessKeyID_DEV) \
-e AWS_SECRET_ACCESS_KEY=$(SecretAccessKey_DEV) \
--volume $(System.DefaultWorkingDirectory)/InfraCostReport:/InfraCostReport/ \
--volume $(System.DefaultWorkingDirectory):/code/ \
infracost/infracost breakdown --sync-usage-file --usage-file /InfraCostReport/infracost-usage.yml \
--path /code/terraform/TerraformPlan.json
```

However, because I'm running InfraCost from the Docker container, the pipeline throws an access denied error. 

```cmd
Error, System.UnauthorizedAccessException: Access to the path '/home/vsts/work/1/s/InfraCostReport/infracost-usage.yml' is denied.

##[error]Access to the path '/home/vsts/work/1/s/InfraCostReport/infracost-usage.yml' is denied.
```

**Comment by AErmie** [comment]
# UPDATE
By adding `--user $(id -u ${USER}):$(id -g ${USER})` to the docker run, I'm able to access the produced infracost-usage.yml file.

Here's the updated full execution command for everyone...

```yaml
docker run --rm \
-e INFRACOST_API_KEY=$(InfraCost_API_KEY) \
-e AWS_ACCESS_KEY_ID=$(AccessKeyID_DEV) \
-e AWS_SECRET_ACCESS_KEY=$(SecretAccessKey_DEV) \
--volume $(System.DefaultWorkingDirectory)/InfraCostReport:/InfraCostReport/ \
--volume $(System.DefaultWorkingDirectory):/code/ \
--user $(id -u ${USER}):$(id -g ${USER}) \
infracost/infracost breakdown --sync-usage-file --usage-file /InfraCostReport/infracost-usage.yml \
--path /code/terraform/TerraformPlan.json
```

**Comment by aliscott** [comment]
Thanks @AErmie. Since the initial panic has been fixed and released in v0.9.9 I'm going to close this issue. If you have any thoughts about how we can improve the usability of the usage file it would be great to hear!
